src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs

```diff
@@ -0,0 +1,280 @@
+    public class SymbolCouldHaveMoreSpecificTypeAnalyzer : DiagnosticAnalyzer
+    {
+        private const string SystemCategory = "System";
+
+        public static readonly DiagnosticDescriptor LocalCouldHaveMoreSpecificTypeDescriptor = new DiagnosticDescriptor(
+            "LocalCouldHaveMoreSpecificType",
+            "Local Could Have More Specific Type",
+            "Local variable {0} could be declared with more specific type {1}.",
+            SystemCategory,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        public static readonly DiagnosticDescriptor FieldCouldHaveMoreSpecificTypeDescriptor = new DiagnosticDescriptor(
+           "FieldCouldHaveMoreSpecificType",
+           "Field Could Have More Specific Type",
+           "Field {0} could be declared with more specific type {1}.",
+           SystemCategory,
+           DiagnosticSeverity.Warning,
+           isEnabledByDefault: true);
+
+        /// <summary>Gets the set of supported diagnostic descriptors from this analyzer.</summary>
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get { return ImmutableArray.Create(LocalCouldHaveMoreSpecificTypeDescriptor, FieldCouldHaveMoreSpecificTypeDescriptor); }
+        }
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(
+                (compilationContext) =>
+                {
+                    Dictionary<IFieldSymbol, HashSet<INamedTypeSymbol>> fieldsSourceTypes = new Dictionary<IFieldSymbol, HashSet<INamedTypeSymbol>>();
+
+                    compilationContext.RegisterOperationBlockStartAction(
+                        (operationBlockContext) =>
+                        {
+                            IMethodSymbol containingMethod = operationBlockContext.OwningSymbol as IMethodSymbol;
+
+                            if (containingMethod != null)
+                            {
+                                Dictionary<ILocalSymbol, HashSet<INamedTypeSymbol>> localsSourceTypes = new Dictionary<ILocalSymbol, HashSet<INamedTypeSymbol>>();
+
+                                // Track explicit assignments.
+                                operationBlockContext.RegisterOperationAction(
+                                   (operationContext) =>
+                                   {
+                                       IAssignmentExpression assignment = (IAssignmentExpression)operationContext.Operation;
+                                       AssignTo(assignment.Target, localsSourceTypes, fieldsSourceTypes, assignment.Value);
+                                   },
+                                   OperationKind.AssignmentExpression,
+                                   OperationKind.CompoundAssignmentExpression,
+                                   OperationKind.IncrementExpression);
```

Comment:
If there was code like Foo(ref field1) would the assignment to field1 be covered by the AssignmentExpression?